src/traction_modem/TractionModem.cxxtest

```diff
@@ -0,0 +1,360 @@
+    }
+
+    int fd_[2]; ///< pipe FDs
+    RxFlow flow_{&g_service}; ///< flow instance under test
+
+    /// Mock packet flow interface.
+    ::testing::StrictMock<MockPacketFlowInterface> mPFI_;
+};
+
+/// Append the expected CRC value.
+/// @param payload packet payload
+/// @param all crc of all bytes
+/// @param even crc of even index bytes
+/// @param odd crc of odd index bytes
+static void append_expected_crc(
+    Defs::Payload *payload, uint16_t all, uint16_t even, uint16_t odd)
+{
+    payload->push_back(all >> 8);
+    payload->push_back(all & 0xFF);
+    payload->push_back(even >> 8);
+    payload->push_back(even & 0xFF);
+    payload->push_back(odd >> 8);
+    payload->push_back(odd & 0xFF);
+}
+
+/// Append the expected CRC value.
+/// @param payload packet payload, including a prepended preamble
+static void append_expected_crc(Defs::Payload *payload)
+{
+    Defs::CRC crc;
+    crc3_crc16_ccitt(payload->data() + sizeof(uint32_t),
+         payload->size() - sizeof(uint32_t), crc.crc);
+    append_expected_crc(payload, crc.all_, crc.even_, crc.odd_);
+}
+
+//
+// TxFlowTest::CreateNoStart
+//
+TEST_F(TxFlowTest, CreateNoStart)
+{
+    auto b = flow_.alloc();
+    b->data()->payload = std::move(Defs::get_fn_set_payload(10, 1));
+    flow_.send(b);
+}
+
+//
+// TxFlowTest::Write
+//
+TEST_F(TxFlowTest, Write)
+{
+    std::string result;
+    std::string expected;
+    init();
+
+    auto b = flow_.alloc();
+    b->data()->payload = std::move(Defs::get_fn_set_payload(10, 1));
+    flow_.send(b);
+    wait_for_main_executor();
+
+    result = recv();
+    expected = "\x41\xd2\xc3\x7a"s "\x01\x01"s "\x00\x06"s
+               "\x00\x00\x00\x0A\x00\x01"s;
+    append_expected_crc(&expected);
+    EXPECT_EQ(expected, result);
+}
+
+//
+// RxFlowTest::CreateNoStart
+//
+TEST_F(RxFlowTest, CreateNoStart)
+{
+}
+
+//
+// RxFlowTest::Read
+//
+TEST_F(RxFlowTest, Read)
+{
+    using ::testing::_;
+
+    std::string data;
+    init();
+    
+    data = "\x41\xd2\xc3\x7a"s "\x01\x01"s "\x00\x06"s
+           "\x00\x00\x00\x0A\x00\x01"s;
+    append_expected_crc(&data);
+
+    // Send a packet without a registered receiver.
+    EXPECT_CALL(mPFI_, send(_, _)).Times(0);
+    send(data);
+    wait_for_main_executor();
+    testing::Mock::VerifyAndClearExpectations(&mPFI_);
+
+    // All the data should have been pulled from the pipe. Since there is no
+    // receiver, it gets thrown on the floor.
+    {
+        uint8_t dummy[4];
+        int result = read(fd_[0], dummy, sizeof(dummy));
+        EXPECT_LT(result, 0);
+        EXPECT_EQ(EAGAIN, errno);
+    }
+
+    // Register a receiver and make sure it gets the next sent message.
+    flow_.set_listener(&mPFI_);
+    EXPECT_CALL(mPFI_, send(testing::Eq(data), UINT_MAX)).Times(1);
+    send(data);
+    wait_for_main_executor();
+    testing::Mock::VerifyAndClearExpectations(&mPFI_);
+
+    // Send a message in pieces in order to exercise some of the other paths.
```

Comment:
We should add another test where we send two packets in one ::write() command and expect both of those to arrive.